src/hotspot/share/gc/shenandoah/shenandoahCollectionSet.inline.hpp

```diff
@@ -40,12 +40,12 @@ bool ShenandoahCollectionSet::is_in(ShenandoahHeapRegion* r) const {
 }
 
 bool ShenandoahCollectionSet::is_in(oop p) const {
-  shenandoah_assert_in_heap(NULL, p);
+  shenandoah_assert_in_heap_or_null(NULL, p);
   return is_in_loc(cast_from_oop<void*>(p));
 }
 
 bool ShenandoahCollectionSet::is_in_loc(void* p) const {
-  assert(_heap->is_in(p), "Must be in the heap");
+  assert(p == NULL || _heap->is_in(p), "Must be in the heap");
   uintx index = ((uintx) p) >> _region_size_bytes_shift;
```

Comment:
Is this right? if heap is not zero-based

Reply:
Yes. The biased cset-map is allocated such that NULL object maps to a special page that always yields false when checking in_cset(NULL). We use that same technique in JIT-compiled code to avoid explicit NULL-checks.